FOSSBilling/FOSSBilling
# Implementing the new extension store

This PR implements the new backend written from scratch in [FOSSBilling/extension-store](https://extensions.fossbilling.org).

## The backend
The backend is a Next.js server with data currently fed statically in a file-based data store. I've kept the API responses identical as much as I could to the old extension store used in 2015 in BoxBilling.

The Next.js app also has some basic frontend done to it temporarily. I couldn't see any reason to prioritize its front end as the majority of the users probably will install their extensions on the admin panel, so I'm not going to work on it any further for some time.

This is an example of the data returned from the "list" API endpoint: https://extensions.fossbilling.org/api/list
You can also return the data of an individual module like so: https://extensions.fossbilling.org/api/extension/serviceproxmox

We can adjust the data here: https://github.com/FOSSBilling/extension-store/blob/main/utils/sample-data.ts

I'm pretty sure the backend will cover our needs until the catalogue grows to a bigger size, I'll consider handling the data in a better way then.

## Integrating it with the FOSSBilling admin panel
The admin panel already had the related layouts and the underlying routes ready, but they needed some overhaul. I've switched it to the ZIP toolkit we're using for our core updater, and from cURL to Guzzle.

I've also updated the tests accordingly, so this PR should be ready to be tested.

![image](https://user-images.githubusercontent.com/35808275/204100588-57b2407d-6a8f-4187-b755-52d076019238.png)

## How to test
- Go to your admin panel
- Extensions -> Overview
- Slide down
- Check the "Modules on the extension store" tab

## Plans for the future

- "Extension details" page inside the admin panel
- Properly writing the new front end for the store
- Version constraints, e.g. modules defining what version of FOSSBilling the module is tested for
- "Active installations" telemetry
- Keeping the old versions in a data store and archiving them
